### ISSUE TITLE
Unable to deploy Longhorn - manager NPE

### ISSUE THREAD

**Issue by billabongrob** [rancher/longhorn#260]
**Environment:**
Using rancher 2.0.7 + kubernetes v1.11.1 + catalog v. 0.3.0
Docker version 17.03.2-ce
RHEL 7.5
Virtual machines

When trying to deploy the catalog application, I'm facing kubelet panics.  

```
E0824 15:58:21.728101 1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:509
/usr/local/go/src/runtime/panic.go:491
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:367
<autogenerated>:1
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/store.go:84
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:261
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:336
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:363
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:151
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:195
<autogenerated>:1
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:227
<autogenerated>:1
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:2337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1117b95]
goroutine 176 [running]:
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x111
panic(0x13f5fe0, 0x2166f30)
/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/rancher/longhorn-manager/k8s/pkg/apis/longhorn/v1alpha1.(*Node).GetNamespace(0x0, 0x0, 0x7f5034589d78)
<autogenerated>:1 +0x5
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1576e00, 0x0, 0x8, 0xc4202ee240, 0x194, 0xc4205acb50)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/store.go:84 +0x132
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc(0x1576e00, 0x0, 0x8, 0xc4205acbb0, 0x112fa63, 0x217d380)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:261 +0x74
github.com/rancher/longhorn-manager/controller.(*NodeController).enqueueNode(0xc4201f25a0, 0x0)
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:336 +0x48
github.com/rancher/longhorn-manager/controller.(*NodeController).enqueueReplica(0xc4201f25a0, 0xc420717b80)
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:363 +0xf5
github.com/rancher/longhorn-manager/controller.NewNodeController.func8(0x15770c0, 0xc420717b80)
/go/src/github.com/rancher/longhorn-manager/controller/node_controller.go:151 +0x45
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc42047e100, 0xc42047e110, 0xc42047e120, 0x15770c0, 0xc420717b80)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:195 +0x49
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc420430200, 0x15770c0, 0xc420717b80)
<autogenerated>:1 +0x62
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd(0xc42047e130, 0x219aa40, 0xc420430200, 0x15770c0, 0xc420717b80)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x6b
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*FilteringResourceEventHandler).OnAdd(0xc420430220, 0x15770c0, 0xc420717b80)
<autogenerated>:1 +0x62
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0xc420088910, 0x0, 0xc420020600)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554 +0x23b
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc4209d3df0, 0x18, 0x8b6350)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203 +0x9c
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548 +0x81
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420585768)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4209d3f68, 0xdf8475800, 0x0, 0x13be701, 0xc4200755c0)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420585768, 0xdf8475800, 0xc4200755c0)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc42041a100)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546 +0x78
github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.(*processorListener).(github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache.run)-fm()
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390 +0x2a
github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc4202de6f0, 0xc4201a6960)
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
/go/src/github.com/rancher/longhorn-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

**Comment by yasker** [comment]
It seems due to this function https://github.com/rancher/longhorn-manager/blob/master/controller/node_controller.go#L358

It doesn't check if GetNode returns nil. I think that's the one caused the manager NPE.

@billabongrob It also means there are some replicas in your system pointed to the nodes no longer existed. Can you show the result of `kubectl -n longhorn-system get lhr -o yaml` and `kubectl -n longhorn-system get lhn`?

**Comment by billabongrob** [comment]
Certainly
```
$ kubectl -n longhorn-system get lhr -o yaml
apiVersion: v1
items: []
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

$ kubectl -n longhorn-system get lhn
NAME             STATUS    AGE       VERSION
srvhourandev01   Ready     73d       v1.11.1
srvhourandev02   Ready     73d       v1.11.1
srvhourandev03   Ready     73d       v1.11.1
```

**Comment by yasker** [comment]
@billabongrob Longhorn managers are still failing? 

And it seems your kubectl version is old. There is a bug before v1.8 or v1.9 that short name was translated incorrectly.

So can you try `kubectl -n longhorn-system get nodes.longhorn.rancher.io -o yaml` instead?

**Comment by billabongrob** [comment]
Still failing.  Sure thing:

```
$ kubectl -n longhorn-system get nodes.longhorn.rancher.io -o yaml
apiVersion: v1
items: []
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

![screen shot 2018-08-24 at 3 53 38 pm](https://user-images.githubusercontent.com/8441943/44607393-ec4a1f80-a7b5-11e8-93e5-2cf7db91cce5.png)

**Comment by yasker** [comment]
Weird... Can you show me the full logs of all the longhorn-manager pods? It's maybe too big to paste here directly.

**Comment by dsargent3220** [comment]
I'm getting the same errors on one of are worker nodes, but not the other.
Long story shorty.  I installed 0.3.0 and it worked fine but then I realized I installed it in the default project,   So I uninstalled it, which did not go well, there where a lot of object left behind and longhorn would not reinstall because the longhorn-system namespace already existed.  After a couple hours of hunting down pods, volumes, pv's, pvc's and deleting them (many of which I had to edit and delete the longhorn finalizer line for it to delete) I finally got the namespace to delete.  
Now when I reinstall in a different project, it works on one node but not the other.

kubectl -n longhorn-system get lhr -o yaml
apiVersion: v1
items:
- apiVersion: longhorn.rancher.io/v1alpha1
  kind: Replica
  metadata:
    creationTimestamp: 2018-08-24T19:30:07Z
    deletionGracePeriodSeconds: 0
    deletionTimestamp: 2018-08-24T22:40:10Z
    finalizers:
    - longhorn.rancher.io
    generation: 2
    labels:
      longhornnode: lpul-k8swrk1tst
      longhornvolume: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
    name: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-r-7fd579f2
    namespace: longhorn-system
    ownerReferences:
    - apiVersion: longhorn.rancher.io/v1alpha1
      kind: longhorn.rancher.io/v1alpha1, Kind=Volume
      name: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
      uid: 1b7826dc-a7d4-11e8-b912-005056b6b782
    resourceVersion: "10146566"
    selfLink: /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/replicas/pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-r-7fd579f2
    uid: 1b7feba9-a7d4-11e8-8412-005056b67263
  spec:
    active: true
    baseImage: ""
    dataPath: /var/lib/rancher/longhorn/replicas/pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-242d0ddd
    desireState: running
    diskID: fd0000000000
    engineImage: rancher/longhorn-engine:v0.3.0
    engineName: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-e-d2216c22
    failedAt: ""
    healthyAt: 2018-08-24T19:30:27Z
    nodeID: lpul-k8swrk1tst
    ownerID: lpul-k8swrk1tst
    restoreFrom: ""
    restoreName: ""
    volumeName: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
    volumeSize: "2147483648"
  status:
    currentImage: rancher/longhorn-engine:v0.3.0
    currentState: running
    ip: 10.42.3.97
    started: true
- apiVersion: longhorn.rancher.io/v1alpha1
  kind: Replica
  metadata:
    creationTimestamp: 2018-08-24T19:30:07Z
    deletionGracePeriodSeconds: 0
    deletionTimestamp: 2018-08-24T22:40:10Z
    finalizers:
    - longhorn.rancher.io
    generation: 2
    labels:
      longhornnode: lpul-k8swrk1tst
      longhornvolume: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
    name: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-r-9ea11a31
    namespace: longhorn-system
    ownerReferences:
    - apiVersion: longhorn.rancher.io/v1alpha1
      kind: longhorn.rancher.io/v1alpha1, Kind=Volume
      name: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
      uid: 1b7826dc-a7d4-11e8-b912-005056b6b782
    resourceVersion: "10146567"
    selfLink: /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/replicas/pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-r-9ea11a31
    uid: 1b7eee6b-a7d4-11e8-8412-005056b67263
  spec:
    active: true
    baseImage: ""
    dataPath: /var/lib/rancher/longhorn/replicas/pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-36f138ad
    desireState: running
    diskID: fd0000000000
    engineImage: rancher/longhorn-engine:v0.3.0
    engineName: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa-e-d2216c22
    failedAt: ""
    healthyAt: 2018-08-24T19:30:27Z
    nodeID: lpul-k8swrk1tst
    ownerID: lpul-k8swrk1tst
    restoreFrom: ""
    restoreName: ""
    volumeName: pvc-19cd17d7-a7d4-11e8-aaae-005056b6e1aa
    volumeSize: "2147483648"
  status:
    currentImage: rancher/longhorn-engine:v0.3.0
    currentState: running
    ip: 10.42.3.96
    started: true
- apiVersion: longhorn.rancher.io/v1alpha1
  kind: Replica
  metadata:
    creationTimestamp: 2018-08-24T19:27:39Z
    deletionGracePeriodSeconds: 0
    deletionTimestamp: 2018-08-24T22:40:10Z
    finalizers:
    - longhorn.rancher.io
    generation: 2
    labels:
      longhornnode: ""
      longhornvolume: test-1
    name: test-1-r-52af16a2
    namespace: longhorn-system
    ownerReferences:
    - apiVersion: longhorn.rancher.io/v1alpha1
      kind: longhorn.rancher.io/v1alpha1, Kind=Volume
      name: test-1
      uid: c379bdb0-a7d3-11e8-b912-005056b6b782
    resourceVersion: "10146568"
    selfLink: /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/replicas/test-1-r-52af16a2
    uid: c37f6c94-a7d3-11e8-8412-005056b67263
  spec:
    active: true
    baseImage: ""
    dataPath: /var/lib/rancher/longhorn/replicas/test-1-54f6384f
    desireState: running
    diskID: fd0000000000
    engineImage: rancher/longhorn-engine:v0.3.0
    engineName: test-1-e-2661da1b
    failedAt: ""
    healthyAt: 2018-08-24T19:28:40Z
    nodeID: ""
    ownerID: lpul-k8swrk1tst
    restoreFrom: ""
    restoreName: ""
    volumeName: test-1
    volumeSize: "5368709120"
  status:
    currentImage: rancher/longhorn-engine:v0.3.0
    currentState: running
    ip: 10.42.3.94
    started: true
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

kubectl -n longhorn-system get lhn
NAME              AGE
lpul-k8swrk2tst   17m

We have two nodes, lpul-k8swrk1tst and lpul-k8swrk2test.    2 worked but 1 did not

**Comment by dsargent3220** [comment]
![image](https://user-images.githubusercontent.com/3372486/44611913-279a1d80-a7b9-11e8-84b5-0de1ec7c39be.png)

**Comment by dsargent3220** [comment]
I got it working.  Thank you Yasker.

**Comment by yasker** [comment]
Thanks @dsargent3220 !

@andreastoermer Are you able to work around the issue by deleting the troubled replicas?

**Comment by yasker** [comment]
Fixed. Will be released with v0.3.1.
